cmd/thanos/bucket.go

```diff
@@ -205,39 +207,41 @@ func registerBucketLs(m map[string]setupFunc, root *kingpin.CmdClause, name stri
 			enc := json.NewEncoder(os.Stdout)
 			enc.SetIndent("", "\t")
 
-			printBlock = func(id ulid.ULID) error {
-				m, err := block.DownloadMeta(ctx, logger, bkt, id)
-				if err != nil {
-					return err
+			printBlock = func(id ulid.ULID, md *metadata.Meta) error {
+				if md == nil {
+					return nil
 				}
-				return enc.Encode(&m)
+				return enc.Encode(&md)
 			}
 		default:
 			tmpl, err := template.New("").Parse(format)
 			if err != nil {
 				return errors.Wrap(err, "invalid template")
 			}
-			printBlock = func(id ulid.ULID) error {
-				m, err := block.DownloadMeta(ctx, logger, bkt, id)
-				if err != nil {
-					return err
+			printBlock = func(id ulid.ULID, md *metadata.Meta) error {
+				if md == nil {
+					return nil
 				}
-
-				if err := tmpl.Execute(os.Stdout, &m); err != nil {
+				if err := tmpl.Execute(os.Stdout, &md); err != nil {
 					return errors.Wrap(err, "execute template")
 				}
 				fmt.Fprintln(os.Stdout, "")
 				return nil
 			}
 		}
 
-		return bkt.Iter(ctx, "", func(name string) error {
-			id, ok := block.IsBlockDir(name)
-			if !ok {
-				return nil
+		mdFetcher := block.NewMetadataFetcher(logger, defaultBlockMDFetchConcurrency, bkt)
```

Comment:
didn't switch to the `download` function because the original logic suggests it requires block id even for partially uploaded blocks.

Reply:
:+1: